### ISSUE TITLE
Update observable with specific event

### ISSUE THREAD

**Issue by hardhub** [knockout/knockout#2204]
I have simple observable... like this:

```JS
var value = ko.observable("initial");
```
and input bound to this value...
also I have a subscription manually subscribed to "value"...

I need to track input value update in this subscription if it was made by user (and then perform some action). But I do not want to perform this action if observable value changed manually from code.
I cannot set any flag and so on... Subscription handler is called much later than code setting observable value.

How to track only changes coming from DOM only?
Is it possible to specify event when observable value set... like this: value("new value", "event-name")?
How to detect inside handler which event is source of update?

**Comment by mbest** [comment]
This seems like a support question, which will get better and quicker responses either on [StackOverflow](stackoverflow.com/questions/tagged/knockout.js) or [Google Groups](http://groups.google.com/group/knockoutjs).

**Comment by hardhub** [comment]
Do you think it is not normal practice to discuss feature requests here? Strange...

**Comment by chrisknoll** [comment]
What's the feature you are asking for?  It sounded like you were asking how to 'switch off' notifications while a block of program executes.

**Comment by mbest** [comment]
@hardhub I don't mind reopening this if you'd like to suggest a feature. Also there's no harm in continuing the conversation here if you like.

**Comment by hardhub** [comment]
> What's the feature you are asking for? It sounded like you were asking how to 'switch off' notifications while a block of program executes.

Nothing specific yet... "Switch of" can be an option... But I need to discuss which solution is better to get desired behaviour.
For now I wrote extension... but it cannot access private member.

```JS
ko.observable.fn.quiet = function (skips) {
    if (skips === undefined) {
        this.notifySubscribers = function () {
        };
    } else {
        this.notifySubscribers = function () {
            var subscriptions = this._subscriptions.change;

            var targets = [];
            for (var index = 0; index < subscriptions.length; index++) {
                var subscription = subscriptions[index];

                var include = true;
                for (var skip = 0; skip < skips.length; skip++) {
                    if (skips[skip] === subscription)
                        include = false;
                }

                if (include === true)
                    targets.push(subscription);
            }

            this._subscriptions.change = targets;

            ko.subscribable.fn.notifySubscribers.apply(this, arguments);

            this._subscriptions.change = subscriptions;
        };
    }

    return this;
};
ko.observable.fn.watch = function () {
    this.notifySubscribers = ko.subscribable.fn.notifySubscribers;

    return this;
};
```


So I can mute for specific subscribers and watch again after changing of observable.

But it is trick for debug version of KO. And it could be good to have something in release to achieve it.

So to clarify the case... We cannot detect who is source of event. For example user types in input and that data should be processed (function passed to subscribe method should be called). And this scenario supported well. But if I need in addition to to restore already ready data in this input I will use the same observable and then that observable will trigger subscription. And we cannot simply set boolean flag... observable notifies after flag restored (I guess because it works async) .

So in original post I wrote above 2 potential solutions:
1) Set observable value with some subscriptions disabled (all subscriptions by default). For example "quiet" method and "watch" to get it back.
2) Set observable value with specified custom event source (not "change") and add some way to detect it inside function passed to subscribe method. So it will be possible to know who is the initiator of current calling and do not make any action inside subscribe callback.

> one option would be to use an observable solely for the user input. You can then subscribe to that observable and pass along the value into your system is a specific way.

It will not help in case described above.
In other words some code do not have the way to not react for changes.
And this is not OK if we need to restore value in UI and then continue tracking of changes.
Have I misunderstood something?

**Comment by brianmhunt** [comment]
Perhaps a simple extender? e.g.

```
ko.extenders.off = function (target, switch) {
  return ko.computed({
    read: target,
    write: (v) => switch() && target(v)
  })
}

var condition = ko.observable(false)
var x = ko.observable().extend({off: condition})
```

It doesn't address the app-specific context where you'd want to turn the observable on/off, but certainly it's a simple solution that may help get you thinking on the right track.

**Comment by mbest** [comment]
Have you looked at this? http://www.knockmeout.net/2011/04/pausing-notifications-in-knockoutjs.html

**Comment by hardhub** [comment]
@brianmhunt 

If I understand this code correctly, if switch is off then target(v) will never be called... and original observable "target" (used for input value binding) will not be updated at all... UI element in DOM will not be updated as well... So we will disable all subscriptions, but we need to update only some of them. It does not work.
> write: (v) => switch() && target(v)

But observable must be updated because it is single way to change UI.
I understand from point of view of FRP is not very good to disable any subscriptions.
So as alternative I proposed to use "source" of update (custom event source).

**Comment by hardhub** [comment]
@mbest 

Maybe I misunderstand something but how can it help?
We need to write to "target" observable to update UI.
Disabling of all subscriptions cannot help because DOM should be notified but custom handler not.

I think in some cases it can be solved by this way:
1) Unsubscribe custom subscription.
2) Update observable with some value to restore DOM value.
3) Subscribe the same subscription again to track changes.

Of course it will look ugly... especially if observable member of array... but that is not all:
If observable is changed multiple times before subscribing again (for example restoring UI by observable set and some other action.. even very quick user input) then custom subscription will not be called (because removed) but it should be: it should react for user input.

**Comment by jrsearles** [comment]
It seems like you need to create a custom handler that triggers a custom event when the DOM update is triggered.

```js
ko.bindingHandlers.textInputExt = {
  init: function (element, valueAccessor) {
    // depending on your needs the "change" event might be more appropriate
    element.addEventListener("input", function () {
      var observable = valueAccessor();
      if (ko.isWritableObservable(observable)) {
        observable(element.value); 
        observable.notifySubscribers(element.value, "userUpdate");
      } else {
        // it gets more complicated...
      }
    });

    ko.computed(function () {
      var value = ko.unwrap(valueAccessor());
      if (value != element.value) {
        element.value = value;
      }
    }, null, { disposeWhenNodeIsRemoved: element });
  }
};
```

Note that this isn't including clean up or cross-browser support. You'd want to reference the original implementation and/or use a DOM library to account for browser quirks.

Then in your consuming code:

```js
yourObservable.subscribe(handler, null, "userUpdate");
```

**Comment by hardhub** [comment]
@jrsearles

I guess this code "element.value = value;" is code for DOM update.. like "ko.selectExtensions.writeValue(element, newValue, allowUnset);" in original code.

So I see 2 issues in suggested behavior:

1) Imagine scenario:
App gets data from server.
Sets it to observable to restore in UI.
Observable changes DOM element.
DOM element fires event.
textInputExt hooks that DOM event.
It fires this: observable.notifySubscribers(element.value, "userUpdate");
Finally this handler will be called: yourObservable.subscribe(handler, null, "userUpdate");

But I should not get it (change was initiated by logic)... I should get notification only if user really types new value.
And as I said previously I cannot control that logic.. cannot set flag.. async updates... notification happens much later after set completed.

Maybe we can set some private var as flag in binding itself and check its status in event listener.
But in that case we still have issue with "multiple updates" (from code and user concurrently).

2) Depending on DOM element I will have to change each binding used in the component (and in the app in common). Original binding code are long and very tricky for wide browser support... So this pretty looks like a good case for feature request.

**Comment by chrisknoll** [comment]
I'm still not getting a sense of a 'feature' you are asking for, this is more like application specific behavior that has some complication to it, so you're going to have to write some  complicated code.

However, I think the other posters have given you the tools you need to solve this.  I think you said it yourself that the key is that you don't want to do logic when the application modifies your model, but only when the user enters a new value.

So why are you trying to subscribe to the model that is populated from the server?  Instead, set up a user input observable that is bound to the input component, and subscribe to THAT object, and perform custom logic, and finish it with writing the value that was entered by the user into your model observable.  I believe you can accomplish this with a writable observable that @brianmhunt suggested, but without the switch. Just read the value from the server-based model and when the value is written, you know it came from user input and perform your custom action. You can even make a parameter to your custom observable take a function which is your handler, so you don't have to duplicate any code, you just need to provide the callback handler on a case-by-case basis.

**Comment by mbest** [comment]
Knockout isn't an eventing system. Knockout deals with values, either plain or computed. So how do you know if the user changed the value? You keep two copies of the value. One is the value from the server and another is the value bound to the UI. At any time, you can compare those values to see whether the user has changed it.

```
ko.computed(function() {
    if (xUser() !== xServer()) {
        // The user has changed the value... do something
    }
});
```

**Comment by hardhub** [comment]
@chrisknoll

You did not understand anything written above.

But I cannot understand what exactly is not clear in this scenario?

> App gets data from server.
> Sets it to observable to restore in UI.
> Observable changes DOM element.
> DOM element fires event.
> textInputExt hooks that DOM event.
> It fires this: observable.notifySubscribers(element.value, "userUpdate");
> Finally this handler will be called: yourObservable.subscribe(handler, null, "userUpdate");

I will try in another way... simple as possible.
You have 2 types of triggers of observable change: a) event from DOM, b) setter of observable like myobservable("new_value")
Also you subscribed to myobservable with subscribe method: myobservable.subscribe(myhandler). It should be called only if it is triggered by DOM. But as it is for now it will be called by change from any trigger: no ability to recognize the trigger.

How can you set value of observable with updating DOM but without executing myhandler?

2 options proposed: 1) choose which subscriptions should be notified on observable update and which not 2) allow to pass custom event source like this myobservable("new_value", "custom_change"), and check inside subscription handler (myhandler)... maybe get it as some parameter.

Does it make any sense now for you?

**Comment by hardhub** [comment]
@mbest 

I do not understand what you mean saying event system... it can be very wide meaning...
Also it is not clear why regular computed if we are talking about observable which have getter and setter.

It is not clear what is xUser and xServer....

I think I need provide some example...

```<input data-bind="value: myobservable" />```

```js
var myobservable = ko.observable("0");

var some_callback = function (value) {
    myobservable(value);
};

myobservable.subscribe(function (value) {
    call_some_function(value); // <-- do not call this if value comes from some_callback
});
```

**Comment by mbest** [comment]
What is it that you are doing in `call_some_function` that you don't want to do with all values?

**Comment by hardhub** [comment]
it could be any reason but in my specific case some_callback and call_some_function are receiving and sending functions... 1st receives some value over network and 2nd sends it... they works with different version of the same data.

but in reality it does not matter.. it can be very long stack... designation of methods is not important.. it is just some asynchronous code... it is only important to not send to "somebody" the same value which I just got from "him"

**Comment by hardhub** [comment]
In synchronous word we can do the following:

```js
var restoring = false;
var myobservable = ko.observable("0");

var some_callback = function (value) {
    restoring = true;
    myobservable(value);
    restoring = false;
};

myobservable.subscribe(function (value) {
    if (restoring === false)
        call_some_function(value); // <-- do not call this if it comes from some_callback
});
```

with knockout we cannot do that.. it is async... 
and due to that it has a lot of sync issues...
I already got many troubles with readiness of components, but it is different story... let's solve this at least )))

**Comment by mbest** [comment]
As I mentioned, you're dealing with values.

```js
var serverValue = "0";
var myobservable = ko.observable("0");

var some_callback = function (value) {
    serverValue = value;
    myobservable(value);
};

myobservable.subscribe(function (value) {
    if (value !== serverValue) {
        serverValue = value; // maybe not update this here, but somewhere once we know the server has the new value
        call_some_function(value);
    }
});
```

**Comment by hardhub** [comment]
I think it is very bad idea to do it so. This code even has bugs.
KO code is async between input update and notifying subscribers, so it can be easy as this:
some_callback: serverValue = 10, input set with 10
some_callback: serverValue = 5, input set with 5
handler:  sends 10 to server (because 10 != 5)  // 10 - 1st value to notify, 5 current serverValue
handler:  not sends 5 to server (because 5 == 5)  // 5 - 2nd value to notify, 5 current serverValue
of course it will be more worse if values will income very often... + user input...

consequences of this bug:
1) we made request to server without any user interaction.
2) wrong value passed to server... there was no user interaction.. so last value should be as passed from server - 5, but really was updated to 10...

I personally do not think it is question of data, appropriate tools to manage subscriptions should be available in knockout... otherwise we do not have a way to correctly react on UI changes in async manner...

**Comment by chrisknoll** [comment]
@hardhub, please consider that you may not have understood my solution.  If you could consider what @mbest is saying as 'knockout is not an venting system', the point I think he's trying to make is that you're looking at knockout from the wrong direction.  It's a data-binding library, not an event notification library.  It goes data-> dom, not dom->data.  I'l try to give you a concrete example based on @brianmhunt 's example:

```
ko.extenders.intercept= function (target, callback) {
  return ko.computed({
    read: target,
    write: function(newValue) { callback(newValue); target(newValue)}
  })
}

var myCallback = function(newValue) {
  console.log("From intercept callback: " + newValue);
}

var pageModel = {
  serverValue = ko.observable(0);
  uiValue = serverValue.extend({intercept: myCallback});
};

ko.applyBindings(pageModel);

```

In HTML:
```
The server value is: <span data-bind="text: serverValue "></span>
The UI value is: <input data-bind="value: uiValue"></input>
```

What should happen here is that the page model's serverValue will be wrapped by the extender 'intercept', and the intercept will respond to changes made via the input field, and nothing subscribes to the serverValue observable (or uiValue for that matter).  Also, this behavior assumes that when the requset to the server updates the serverValue, the textInput value will also change (via the read on the computedObservable) but read doesn't trigger the callback, only write does.

Hopefully what you can take away from this is that you shouldn't be hooking into dom events in this case. You're getting in the way of knockout.  Try to use the constructs available to you in the framework, or don't use it all together.  Please don't respond telling me I don't understand the problem: leveraging dom events isn't what you need to solve this, IMO.  If you have some hard dependency on dom events in your context, maybe you should rethink your design.

**Comment by hardhub** [comment]
```
The server value is: <span data-bind="text: serverValue "></span>
The UI value is: <input data-bind="value: uiValue"></input>
```

This is something wrong... it not my case... Let's say we need only this:
`<input data-bind="value: uiValue"></input>`

So to update this value to new one we need to set uiValue observable.
`uiValue("1");`

internally it will be `function(newValue) { callback(newValue); target(newValue)
`
it will call callback with ajax request inside (console.log in your example)
but it is strange logic... I do not need to send this data back to server...

I wrote simple example of what I need to achieve..
I still do not see the solution! Just talks about how MVVM is cool.

**Comment by chrisknoll** [comment]
The html example was just to demonstrate reading from the serverValue, and binding the UI to a model value (which when you update via the ui, the uiValue is an observable intercept of the serverValue observable.  If you only want to display the UI element, fine.

You would never call 
```
uiValue("1");
```

The whole point fo the uiValue model property is to distinguish updates directly to serverValue (from your ajax call) vs. the user input.  the read-value of uiValue will automatically reflrect the value of serverValue() when you updated it asynchronously.

How about we take your workflow from prior post and apply it to my solution:

> App gets data from server.
> Sets it to observable to restore in UI.
> Observable changes DOM element.
> DOM element fires event.

Now, what i'm proposing doesn't deal with DOM events, but, what would happen is App gets data from server (assuming this is asynchronous) and updates serverValue().  The UI will reflect the server's value due to the 'intercept' observable extension is to return the wrapped serverValue() in 'read'.  Let's say the app gets data again, serverValue() is updated, the UI reflectsthe new value.  Now, the user enters a new value in the input field.  Now it is going to perform the custom callback (you specified that you want to do some logic only when the user enters a new value, so this is where you put that 'user-input-only logic').  I've coded the observable extension to do the custom logic and set the wrapped observable (serverValue in this case) to the new value.  I'm assuming you want that.  If you don't want to do that, then don't update serverValue() and just depend on the next async update of serverValue() to update it for you.

Under what use case of the flow of data from the  server vs. the flow of the UI does this not work for you? Can you give me steps as you've done in the other replies?

**Comment by chrisknoll** [comment]
I think there's something else that you might be misunderstanding:

> This is something wrong... it not my case... Let's say we need only this:
> <input data-bind="value: uiValue"></input>
> 

You seem to be saying that you can only manipulate values that are rendered in HTML.  That's very 'dom-oriented' thinking.  the serverValue is in your model and you can have it store state independently of if it is rendered on the screen or not.

**Comment by hardhub** [comment]
> You would never call

OK, what should be called then to pass data came from server?

serverValue("1"); 

Then uiValue will be updated as subscription... DOM will be updated from uiValue. 

If DOM is updated then uiValue updated, then callback fired and serverValue updated once again (but nothing really happens - value is the same)...

So it looks working... nice )) at least you understand now what I need ))

But I afaraid it has the same issue as solution of @mbest 

P.S.
> You seem to be saying that you can only manipulate values that are rendered in HTML. That's very 'dom-oriented' thinking.
It is just wrong assumption...

**Comment by mbest** [comment]
1. By default, Knockout works synchronously, meaning that when you update an observable, subscribers are called immediately. If you're seeing asynchronous behavior, you're probably using [deferred updates](http://knockoutjs.com/documentation/deferred-updates.html) or the [rate-limit extender](http://knockoutjs.com/documentation/rateLimit-observable.html). In those cases, subscribers will be called once with the latest value even if there were multiple intermediate values.

2. Even if your subscription might get called with a stale value (somehow), you can always read the latest value from the observable: `if (myobservable() !== serverValue)`.

**Comment by hardhub** [comment]
Regarding issue... let's say it is not 5 seconds it is some milliseconds...

Timer callbacks can be processed in this sequence:
1st callback initial set of serverValue
2nd callback initial set of serverValue
1st callback second set of serverValue  "target(newValue)"
2nd callback second set of serverValue "target(newValue)"

Infinite loop if I understand right?

Significantly tired.. need more clear brain to analyze... but I guess it will be infinite loop.

**Comment by chrisknoll** [comment]
There's no infinite loop... setting serverValue to a value does not trigger the callback.  Not sure why you are missing this. The only time a call back is fired is when the computed observable is written to:

```
  uiValue = serverValue.extend({intercept: myCallback});
```

Why do you have '1st callback second set of serverValue "target(newValue)"'?  When the input field gets a new value (from user input), then and ONLY then does it call the call back, and only once.  The only problem i see is if you do have 5 miliseconds, you'll never get a chance to type in any input, and if that's the case, then you should just have a completely independent ui element that is detached from the server value, that writes to an intermediate model property, and then after focus is lost, it submits back the new value to the server so that it gets refreshed in the serverValue.  Not sure if a 5milisecond polling time makes any sense, but  the point is taken that you dont' want to the ui to refresh from the server variable if the user is currently editing.  Therefore, the proper pattern for this interface is to have a read-only form of the server variable (such as in a span) and then an update-mode of the input when the user clicks (which does not read from serverValue,(), it just takes the input and submits the update to the server, and then replaces the read-only form of the ui element when the user leaves focus.

See http://knockoutjs.com/documentation/hasfocus-binding.html example 2 for the 'click to edit' example.

**Comment by hardhub** [comment]
> By default, Knockout works synchronously, meaning that when you update an observable, subscribers are called immediately.

Immediately but not synchronously... at least it is what I have inside real app.
```js
var myobservable = ko.observable("0");

var some_callback = function (value) {
    // 1st point
    myobservable(value);
    // 2nd point
};

myobservable.subscribe(function (value) {
    // 3rd point
    call_some_function(value); // <-- do not call this if value comes from some_callback
});
```

I guess this sample can be used as example for our current discussion.. right?

> deferred updates or the rate-limit extender.

no... in some cases numeric extender.. but with and without extender the behaviour is the same...
but numeric extender...
```js
ko.extenders.numeric = function (target, options) {
    // Create a writeable computed observable to intercept writes to our observable.
    var result = ko.pureComputed({
        read: target,
        write: function (newValue) {
            var current = target();
            var precision = options.precision;
            var minimum = options.minimum;
            var maximum = options.maximum;

            var valueToWrite = null;
            if (newValue != null) {
                var roundingMultiplier = Math.pow(10, precision);
                var newValueAsNum = isNaN(newValue) ? 0 : parseFloat(+newValue);
                valueToWrite = (Math.round(newValueAsNum * roundingMultiplier) / roundingMultiplier);

                if (minimum != null && minimum > valueToWrite)
                    valueToWrite = minimum;

                if (maximum != null && maximum < valueToWrite)
                    valueToWrite = maximum;
            }

            if (current === valueToWrite && newValue !== valueToWrite) {
                target(newValue);
                target(valueToWrite);
            }
            else {
                target(valueToWrite);
            }
        }
    }).extend({ notify: 'always' });

    // Initialize with current value to make sure it is rounded appropriately.
    result(target());

    // Return the new computed observable.
    return result;
};
```

**Comment by hardhub** [comment]
> Why do you have '1st callback second set of serverValue "target(newValue)"'? 

what do you mean?

**Comment by chrisknoll** [comment]
I mean setting serverValue() to a new value does not result in an infinite loop. Why would it?

**Comment by mbest** [comment]
> Immediately but not synchronously... at least it is what I have inside real app.

This isn't possible in Javascript (no multi-threading). Regardless, my second point remains. You can always read the latest value from the observable. It will never by stale.

**Comment by chrisknoll** [comment]
@hardhub , I'm re-reading part of what you wrote here:

> OK, what should be called then to pass data came from server?
> 
> serverValue("1");
> 
> Then uiValue will be updated as subscription... DOM will be updated from uiValue.
> 
> If DOM is updated then uiValue updated, then callback fired and serverValue updated once again (but nothing really happens - value is the same)...
> 

Where you 'got this wrong' is "if the DOM is updated, the uiValue is updated".  You (again) are thinking in the dom -> data mode, not data-> dom mode. The uiValue is not some DOM value.  The DOM is updated when the value of the bound observable detects a change.  When serverValue("1") is updated, the reason why the ui will update is because the computed (created by the intercept extender) will tell the DOM (via the 'value' databinding which is listing to the 'read' of the computed) that there is a new value to assign to the DOM.  This does not trigger a callback because the uiValue (the computed) was not written to.  The only way to get the callback to trigger is to write a value like this:
```
uiValue("1"); 
```
Which i said earlier you would not do, because you only want to write to that observable as a result of UI input, which is why you have this HTML data-binding:

```
<input data-bind="value: uiValue"></input>
```

You want this to be the only code-path to updating the uiValue observable.

You are still left with the problem that if you are in the middle of editing the value that uiValue is bound to, and behind the scenes an update to serverValue() comes through, it will overwrite your UI element while you are editing.  What you need to do in this case is when editing begins (such when element gains focus), you need to take a copy of the current serverValue, and place that in a completely independent observable that will be bound to a textbox .  After the value is updated (which you can use the intercept extender for this) you want to write the value back to the serverValue, or you want to trigger an ajax call to send the new value to the server, or maybe both... your usecase is not clear on this.  

If the serverValue is refreshed based on a user action (such as a reload button) then at least that way you can be sure that serverValues won't change while you're editing.  But if it's completely random (ie you have a websocket which triggers the UI to refresh on-demand, or you have a fixed polling interval) then you will need to work with a copy of the serverValue while editing and then write it back (or make an ajax call) after the write executes.

**Comment by hardhub** [comment]
@mbest 
> This isn't possible in Javascript (no multi-threading). Regardless, my second point remains. You can always read the latest value from the observable. It will never be stale.

deferred was used globally.. so it was possible...

`ko.options.deferUpdates = true;`

OK, so what should I do in this case?
Is it good idea to disable deferred updates?
How to control subscriptions if defered enabled?

**Comment by mbest** [comment]
Thanks for the additional information. I still think comparing the values is the most robust solution.

That being said, you mentioned not being able to dispose the subscription temporarily; that it was called anyway. A fix for that problem is already checked in (#2163), but you'll have to wait for the next version of Knockout or use a custom build to get it.
